homeassistant/components/mqtt/tag.py

```diff
@@ -50,129 +42,95 @@
 )
 
 
-async def async_setup_entry(hass, config_entry):
-    """Set up MQTT tag scan dynamically through MQTT discovery."""
-    setup = functools.partial(async_setup_tag, hass, config_entry=config_entry)
-    await async_setup_entry_helper(hass, "tag", setup, PLATFORM_SCHEMA)
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+    """Set up MQTT device automation dynamically through MQTT discovery."""
+
+    setup = functools.partial(_async_setup_tag, hass, config_entry=config_entry)
+    await async_setup_entry_helper(hass, TAG, setup, PLATFORM_SCHEMA)
 
 
-async def async_setup_tag(hass, config, config_entry, discovery_data):
+async def _async_setup_tag(
+    hass: HomeAssistant,
+    config: ConfigType,
+    config_entry: ConfigEntry,
+    discovery_data: dict,
+) -> None:
     """Set up the MQTT tag scanner."""
     discovery_hash = discovery_data[ATTR_DISCOVERY_HASH]
     discovery_id = discovery_hash[1]
 
-    device_id = None
-    if CONF_DEVICE in config:
-        _update_device(hass, config_entry, config)
-
-        device_registry = dr.async_get(hass)
-        device = device_registry.async_get_device(
-            {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
-            {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
-        )
-
-        if device is None:
-            return
-        device_id = device.id
-
-        if TAGS not in hass.data:
-            hass.data[TAGS] = {}
-        if device_id not in hass.data[TAGS]:
-            hass.data[TAGS][device_id] = {}
+    device_id = update_device(hass, config_entry, config)
+    hass.data.setdefault(TAGS, {})
+    if device_id not in hass.data[TAGS]:
+        hass.data[TAGS][device_id] = {}
 
     tag_scanner = MQTTTagScanner(
         hass,
         config,
-        device_id,
+        cast(str, device_id),
```

Comment:
This cast seems a bit strange.
According to the config schema, it's possible to setup an MQTT tag without a device, in that case the `device_id` parameter to `MQTTTagScanner` should be typed as `str | None`.